// @neo4j/graphql/dist/translate/queryAST/ast/operations/ReadOperation.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOperation = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const utils_1 = require("../../../../utils/utils");
const context_has_target_1 = require("../../utils/context-has-target");
const create_node_from_entity_1 = require("../../utils/create-node-from-entity");
const CypherAttributeField_1 = require("../fields/attribute-fields/CypherAttributeField");
const CypherPropertySort_1 = require("../sort/CypherPropertySort");
const operations_1 = require("./operations");
const wrap_subquery_in_calls_1 = require("../../utils/wrap-subquery-in-calls");
class ReadOperation extends operations_1.Operation {
    constructor({ target, relationship, directed, }) {
        super();
        this.fields = [];
        this.filters = [];
        this.authFilters = [];
        this.sortFields = [];
        this.target = target;
        this.directed = directed ?? true;
        this.relationship = relationship;
    }
    setFields(fields) {
        this.fields = fields;
    }
    addSort(...sort) {
        this.sortFields.push(...sort);
    }
    addPagination(pagination) {
        this.pagination = pagination;
    }
    setFilters(filters) {
        this.filters = filters;
    }
    addAuthFilters(...filter) {
        this.authFilters.push(...filter);
    }
    getAuthFilterSubqueries(context) {
        return this.authFilters.flatMap((f) => f.getSubqueries(context));
    }
    getAuthFilterPredicate(context) {
        return (0, utils_1.filterTruthy)(this.authFilters.map((f) => f.getPredicate(context)));
    }
    transpileNestedRelationship(entity, { context }) {
        const isCreateSelection = context.env.topLevelOperationName === "CREATE";
        //TODO: dupe from transpile
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        const relVar = (0, create_node_from_entity_1.createRelationshipFromEntity)(entity);
        const targetNode = (0, create_node_from_entity_1.createNodeFromEntity)(entity.target, context.neo4jGraphQLContext);
        const relDirection = entity.getCypherDirection(this.directed);
        const pattern = new cypher_builder_1.default.Pattern(context.target)
            .withoutLabels()
            .related(relVar)
            .withDirection(relDirection)
            .to(targetNode);
        const nestedContext = context.push({ target: targetNode, relationship: relVar });
        const filterPredicates = this.getPredicates(nestedContext);
        const authFilterSubqueries = this.getAuthFilterSubqueries(nestedContext).map((sq) => new cypher_builder_1.default.Call(sq).innerWith(targetNode));
        const authFiltersPredicate = this.getAuthFilterPredicate(nestedContext);
        const { preSelection, selectionClause: matchClause } = this.getSelectionClauses(nestedContext, pattern);
        const wherePredicate = cypher_builder_1.default.and(filterPredicates, ...authFiltersPredicate);
        let withWhere;
        let filterSubqueryWith;
        // This weird condition is just for cypher compatibility
        const shouldAddWithForAuth = authFiltersPredicate.length > 0;
        if (authFilterSubqueries.length > 0 || shouldAddWithForAuth) {
            if (!isCreateSelection || authFilterSubqueries.length) {
                // for creates auth filters sometimes use variables from the subquery
                filterSubqueryWith = new cypher_builder_1.default.With("*");
            }
        }
        if (wherePredicate) {
            if (filterSubqueryWith) {
                filterSubqueryWith.where(wherePredicate); // TODO: should this only be for aggregation filters?
            }
            else {
                matchClause.where(wherePredicate);
            }
        }
        const cypherFieldSubqueries = this.getCypherFieldsSubqueries(nestedContext);
        const subqueries = cypher_builder_1.default.concat(...this.getFieldsSubqueries(nestedContext), ...cypherFieldSubqueries);
        const sortSubqueries = (0, wrap_subquery_in_calls_1.wrapSubqueriesInCypherCalls)(nestedContext, this.sortFields, [targetNode]);
        const ret = this.getProjectionClause(nestedContext, context.returnVariable, entity.isList);
        const clause = cypher_builder_1.default.concat(...preSelection, matchClause, ...authFilterSubqueries, filterSubqueryWith, withWhere, subqueries, ...sortSubqueries, ret);
        return {
            clauses: [clause],
            projectionExpr: context.returnVariable,
        };
    }
    getProjectionClause(context, returnVariable, isArray) {
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        const projection = this.getProjectionMap(context);
        let aggregationExpr = cypher_builder_1.default.collect(context.target);
        if (!isArray) {
            aggregationExpr = cypher_builder_1.default.head(aggregationExpr);
        }
        const withClause = new cypher_builder_1.default.With([projection, context.target]);
        if (this.sortFields.length > 0 || this.pagination) {
            this.addSortToClause(context, context.target, withClause);
        }
        return withClause.return([aggregationExpr, returnVariable]);
    }
    getPredicates(queryASTContext) {
        return cypher_builder_1.default.and(...[...this.filters].map((f) => f.getPredicate(queryASTContext)));
    }
    getSelectionClauses(context, node) {
        let matchClause = new cypher_builder_1.default.Match(node);
        let extraMatches = this.getChildren().flatMap((f) => {
            return f.getSelection(context);
        });
        if (extraMatches.length > 0) {
            extraMatches = [matchClause, ...extraMatches];
            matchClause = new cypher_builder_1.default.With("*");
        }
        return {
            preSelection: extraMatches,
            selectionClause: matchClause,
        };
    }
    transpile({ context }) {
        if (this.relationship) {
            return this.transpileNestedRelationship(this.relationship, {
                context,
            });
        }
        const isCreateSelection = context.env.topLevelOperationName === "CREATE";
        const node = (0, create_node_from_entity_1.createNodeFromEntity)(this.target, context.neo4jGraphQLContext, this.nodeAlias);
        const filterSubqueries = (0, wrap_subquery_in_calls_1.wrapSubqueriesInCypherCalls)(context, this.filters, [node]);
        const filterPredicates = this.getPredicates(context);
        const authFilterSubqueries = this.getAuthFilterSubqueries(context).map((sq) => new cypher_builder_1.default.Call(sq).innerWith(node));
        const fieldSubqueries = this.getFieldsSubqueries(context);
        const cypherFieldSubqueries = this.getCypherFieldsSubqueries(context);
        const sortSubqueries = (0, wrap_subquery_in_calls_1.wrapSubqueriesInCypherCalls)(context, this.sortFields, [node]);
        const subqueries = cypher_builder_1.default.concat(...fieldSubqueries);
        const authFiltersPredicate = this.getAuthFilterPredicate(context);
        const ret = this.getReturnStatement(context, context.returnVariable);
        const { preSelection, selectionClause: matchClause } = this.getSelectionClauses(context, node);
        let filterSubqueryWith;
        let filterSubqueriesClause = undefined;
        // This weird condition is just for cypher compatibility
        const shouldAddWithForAuth = authFiltersPredicate.length > 0;
        if (filterSubqueries.length > 0 || shouldAddWithForAuth) {
            filterSubqueriesClause = cypher_builder_1.default.concat(...filterSubqueries);
            if (!isCreateSelection) {
                filterSubqueryWith = new cypher_builder_1.default.With("*");
            }
        }
        const wherePredicate = isCreateSelection
            ? filterPredicates
            : cypher_builder_1.default.and(filterPredicates, ...authFiltersPredicate);
        if (wherePredicate) {
            if (filterSubqueryWith) {
                filterSubqueryWith.where(wherePredicate); // TODO: should this only be for aggregation filters?
            }
            else {
                matchClause.where(wherePredicate);
            }
        }
        let sortClause;
        if (this.sortFields.length > 0 || this.pagination) {
            sortClause = new cypher_builder_1.default.With("*");
            this.addSortToClause(context, node, sortClause);
        }
        const sortBlock = cypher_builder_1.default.concat(...sortSubqueries, sortClause);
        const sortAndLimitBlock = this.hasCypherSort()
            ? cypher_builder_1.default.concat(...cypherFieldSubqueries, sortBlock)
            : cypher_builder_1.default.concat(sortBlock, ...cypherFieldSubqueries);
        let clause;
        // Top-level read part of a mutation does not contains the MATCH clause as is implicit in the mutation.
        if (isCreateSelection) {
            clause = cypher_builder_1.default.concat(filterSubqueriesClause, filterSubqueryWith, sortAndLimitBlock, subqueries, ret);
        }
        else {
            clause = cypher_builder_1.default.concat(...preSelection, matchClause, ...authFilterSubqueries, filterSubqueriesClause, filterSubqueryWith, sortAndLimitBlock, subqueries, ret);
        }
        return {
            clauses: [clause],
            projectionExpr: context.returnVariable,
        };
    }
    getReturnStatement(context, returnVariable) {
        const projection = this.getProjectionMap(context);
        if (context.shouldCollect) {
            return new cypher_builder_1.default.Return([cypher_builder_1.default.collect(projection), returnVariable]);
        }
        return new cypher_builder_1.default.Return([projection, returnVariable]);
    }
    hasCypherSort() {
        return this.sortFields.some((s) => s instanceof CypherPropertySort_1.CypherPropertySort);
    }
    getChildren() {
        return (0, utils_1.filterTruthy)([
            ...this.filters,
            ...this.authFilters,
            ...this.fields,
            this.pagination,
            ...this.sortFields,
        ]);
    }
    getFieldsSubqueries(context) {
        const nonCypherFields = this.fields.filter((f) => !(f instanceof CypherAttributeField_1.CypherAttributeField));
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        return (0, wrap_subquery_in_calls_1.wrapSubqueriesInCypherCalls)(context, nonCypherFields, [context.target]);
    }
    getCypherFieldsSubqueries(context) {
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        return (0, wrap_subquery_in_calls_1.wrapSubqueriesInCypherCalls)(context, this.getCypherFields(), [context.target]);
    }
    getCypherFields() {
        return this.fields.filter((f) => {
            return f instanceof CypherAttributeField_1.CypherAttributeField;
        });
    }
    getProjectionMap(context) {
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        const projectionFields = this.fields.map((f) => f.getProjectionField(context.target));
        const sortProjectionFields = this.sortFields.map((f) => f.getProjectionField(context));
        const uniqueProjectionFields = Array.from(new Set([...projectionFields, ...sortProjectionFields])); // TODO remove duplicates with alias
        const stringFields = [];
        let otherFields = {};
        for (const field of uniqueProjectionFields) {
            if (typeof field === "string")
                stringFields.push(field);
            else {
                otherFields = { ...otherFields, ...field };
            }
        }
        return new cypher_builder_1.default.MapProjection(context.target, stringFields, otherFields);
    }
    addSortToClause(context, node, clause) {
        const isNested = Boolean(context.source); // This is to keep Cypher compatibility
        const orderByFields = this.sortFields.flatMap((f) => f.getSortFields(context, node, !isNested));
        const pagination = this.pagination ? this.pagination.getPagination() : undefined;
        clause.orderBy(...orderByFields);
        if (pagination?.skip) {
            clause.skip(pagination.skip);
        }
        if (pagination?.limit) {
            clause.limit(pagination.limit);
        }
    }
}
exports.ReadOperation = ReadOperation;
