// @neo4j/graphql/dist/schema-model/entity/model-adapters/UnionEntityOperations.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnionEntityOperations = void 0;
const upper_first_1 = require("../../../utils/upper-first");
class UnionEntityOperations {
    constructor(unionEntityAdapter) {
        this.unionEntityAdapter = unionEntityAdapter;
        this.pascalCasePlural = (0, upper_first_1.upperFirst)(this.unionEntityAdapter.plural);
        this.pascalCaseSingular = (0, upper_first_1.upperFirst)(this.unionEntityAdapter.singular);
    }
    get whereInputTypeName() {
        return `${this.unionEntityAdapter.name}Where`;
    }
    get subscriptionEventPayloadTypeName() {
        return `${this.unionEntityAdapter.name}EventPayload`;
    }
    get rootTypeFieldNames() {
        return {
            // create: `create${this.pascalCasePlural}`,
            read: this.unionEntityAdapter.plural,
            // update: `update${this.pascalCasePlural}`,
            // delete: `delete${this.pascalCasePlural}`,
            // aggregate: `${this.unionEntityAdapter.plural}Aggregate`,
            // subscribe: {
            //     created: `${this.unionEntityAdapter.singular}Created`,
            //     updated: `${this.unionEntityAdapter.singular}Updated`,
            //     deleted: `${this.unionEntityAdapter.singular}Deleted`,
            //     relationship_deleted: `${this.unionEntityAdapter.singular}RelationshipDeleted`,
            //     relationship_created: `${this.unionEntityAdapter.singular}RelationshipCreated`,
            // },
        };
    }
}
exports.UnionEntityOperations = UnionEntityOperations;
